src/dataframerow/utils.jl

```diff
@@ -145,50 +146,75 @@ function row_group_slots(cols::Tuple{Vararg{AbstractVector}},
     return ngroups, rhashes, gslots, false
 end
 
-function row_group_slots(cols::Tuple{CategoricalVector},
+function row_group_slots(cols::NTuple{N,<:CategoricalVector},
                          hash::Val{false},
                          groups::Union{Vector{Int}, Nothing} = nothing,
-                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool}
-    col = cols[1]
-    @assert groups === nothing || length(groups) == length(col)
+                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool} where N
+    # Computing neither hashes nor groups isn't very useful,
+    # and this method needs to allocate a groups vector anyway
+    @assert groups !== nothing && all(col -> length(col) == length(groups), cols)
 
     # If missings are to be skipped, they will all go to group 1,
     # which will be removed by group_rows
-    ngroups = length(levels(col)) + (eltype(col) >: Missing)
+    ngroupstup = map(cols) do c
+        length(levels(c)) + (!skipmissing && eltype(c) >: Missing)
+    end
+    ngroups = prod(ngroupstup) + skipmissing
+
+    # Fall back to hashing if there would be too many empty combinations.
+    # The first check ensures the computation of ngroups did not overflow.
+    # The rationale for the 2 threshold is that while the fallback method is always slower,
+    # it allocates a hash table of size length(groups) instead of the remap vector
+    # of size ngroups (i.e. the number of possible combinations) in this method:
+    # so it makes sense to allocate more memory for better performance,
+    # but it needs to remain reasonable compared with the size of the data frame.
+    if prod(Int128.(ngroupstup)) > typemax(Int) || ngroups > 2 * length(groups)
+        return invoke(row_group_slots,
+                      Tuple{Tuple{Vararg{AbstractVector}}, Val,
+                            Union{Vector{Int}, Nothing}, Bool},
+                      cols, hash, groups, skipmissing)
+    end
 
-    if groups !== nothing
+    seen = fill(false, ngroups)
```

Comment:
Have you checked if it would it be better to use `BitSet` here?.

Reply:
OK. I've also tried it, and it's much slower. That sounds logical since a boolean indicator is really the simplest information we need to store here.

Reply:
OK given the discussion above